File: me_daily/2025/11/17/lc_1178.go
package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
	"strconv"
	"strings"
)

var reader = bufio.NewReader(os.Stdin)
var writer = bufio.NewWriter(os.Stdout)

func readLine() string {
	s, _ := reader.ReadString('\n')
	return strings.TrimSpace(s)
}

func readInts() []int {
	fields := strings.Fields(readLine())
	res := make([]int, len(fields))
	for i, f := range fields {
		res[i], _ = strconv.Atoi(f)
	}
	return res
}

func toMask(s string) int {
	m := 0
	for _, c := range s {
		m |= 1 << (c - 'a')
	}
	return m
}

func main() {
	defer writer.Flush()

	head := readInts()
	wl, pl := head[0], head[1]

	// 读入 wl 个单词（同一行）
	wordFields := strings.Fields(readLine())
	words := wordFields[:wl]

	// 读入 pl 行 puzzles
	puzzleFields := strings.Fields(readLine())
	puzzles := puzzleFields[:pl]

	// 频率表：mask -> count
	freq := make(map[int]int)

	for _, w := range words {
		m := toMask(w)
		// 字母数超过7的不可能被 puzzle 覆盖
		if bits := bits.OnesCount(uint(m)); bits <= 7 {
			freq[m]++
		}
	}

	ans := make([]int, pl)

	for i, p := range puzzles {
		pMask := toMask(p)
		firstBit := 1 << (p[0] - 'a')
		rest := pMask ^ firstBit

		cnt := 0
		for sub := rest; ; sub = (sub - 1) & rest {
			cnt += freq[sub|firstBit]
			if sub == 0 {
				break
			}
		}

		ans[i] = cnt
	}

	for i, v := range ans {
		if i > 0 {
			fmt.Fprint(writer, " ")
		}
		fmt.Fprint(writer, v)
	}
	fmt.Fprintln(writer)
}

File: me_daily/2025/11/17/lc_1178.py
from collections import defaultdict
from typing import List
import sys
from collections import Counter
input = lambda: sys.stdin.readline().strip()
II = lambda: int(input())
LII = lambda: list(map(int, input().split()))
SI = lambda: input()
LSI = lambda: input().split()


wl, pl = LII()

words = [SI() for _ in range(wl)]
puzzles = [SI() for _ in range(pl)]

def to_mask(s):
    m = 0
    for c in s:
        m |= 1 << (ord(c) - ord('a'))
    return m


freq = Counter()

for w in words:
    m = to_mask(w)
    if m.bit_count() <= 7: freq[m] += 1

ans = []
for p in puzzles:
    p_mask = to_mask(p)
    first_bit = 1 << (ord(p[0]) - ord('a'))
    rest = first_bit ^ p_mask

    sub = rest
    cnt = 0
    while sub:
        cnt += freq[sub | first_bit]
        sub = (sub - 1) & rest
    
    cnt += freq[first_bit]
    ans.append(cnt)
print(*ans)